### PR TITLE
Remove pages-3.1 feature from faces-4.0 feature

### PR DESCRIPTION
**Pull request by pnicolucci**
for #20805

### CODE REVIEW

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by pnicolucci** [comment]
#libby #build

File: dev/com.ibm.websphere.appserver.features/visibility/public/faces-4.0/io.openliberty.faces-4.0.feature [hunk 1]
```diff
@@ -37,19 +37,18 @@ IBM-ShortName: faces-4.0
 WLP-AlsoKnownAs: jsf-4.0
 Subsystem-Name: Jakarta Server Faces 4.0
 -features=io.openliberty.jakarta.websocket-2.1, \
-  io.openliberty.servlet.api-6.0, \
+  com.ibm.websphere.appserver.servlet-6.0, \
   io.openliberty.jakarta.xmlBinding-4.0, \
```
Comment (pnicolucci) [inline dev/com.ibm.websphere.appserver.features/visibility/public/faces-4.0/io.openliberty.faces-4.0.feature:41]: The pages-3.1 feature enabled the servlet feature and the expression language feature which is why we need to add them directly in the faces-4.0 feature now.

**Comment by pnicolucci** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by pnicolucci** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by pnicolucci** [comment]
#libby #build

**Comment by pnicolucci** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

File: dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/ee/WASMyFacesContainerInitializer.java [hunk 2]
```diff
@@ -83,6 +84,23 @@ public void onStartup(Set<Class<?>> clazzes, ServletContext servletContext) thro
...
 
             log.log(Level.INFO, "Added StartupServletContextListener to the servlet context");
+
+            /*
+             * In previous version of Faces/JSF this was done in the AbstractJSPExtensionFactory.createExtensionProcessor.
+             * In Faces 4.0 the Pages/JSP feature is no longer enabled as Pages/JSP support was removed from the Faces 4.0 Specification.
+             * If an AnnotationHandlerManager is not added here then when the WASCDIAnnotationInjectionProvider tries to get an instance
+             * of an AnnotationHelperManager null is returned.
+             *
+             * The JSFExtensionFactory will do this for applications that define a FacesServlet and we'll create the
```
Comment (volosied) [inline dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/ee/WASMyFacesContainerInitializer.java:94]: Shouldn't the comment say WASMyFacesContainerInitializer instead?

File: dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/extprocessor/JSFExtensionFactory.java [hunk 3]
```diff
@@ -138,6 +138,22 @@ public ExtensionProcessor createExtensionProcessor(IServletContext webapp) throw
         //set a context attribute for later queries on which JSF impl is in use.
         webapp.setAttribute(JSFConstants.JSF_IMPL_ENABLED_PARAM, jsfImplEnabled);
 
+        /*
+         * In previous version of Faces/JSF this was done in the AbstractJSPExtensionFactory.createExtensionProcessor.
+         * In Faces 4.0 the Pages/JSP feature is no longer enabled as Pages/JSP support was removed from the Faces 4.0 Specification.
+         * If an AnnotationHandlerManager is not added here then when the WASCDIAnnotationInjectionProvider tries to get an instance
+         * of an AnnotationHelperManager null is returned.
+         *
+         * The WASMyFacesContainerInitializer will do this for applications that have a FacesServlet defined dynamically
```
Comment (volosied) [inline dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/extprocessor/JSFExtensionFactory.java:147]: Shouldn't the comment say JSFExtensionFactory instead?

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/TestJSFEL.war/src/com/ibm/ws/jsf22/el/components/CustomComponent.java [hunk 4]
```diff
@@ -108,42 +156,72 @@ private ELResolver[] getELResolvers(ELResolver elResolver, ResponseWriter respon
...
-                    if (!resolvers[i].getClass().toString().equals(expectedResolvers.get(i))) {
+                    if (!resolver.equals(expectedResolvers.get(i))) {
                         return false;
                     }
                 }
             }
+
             if (elResolversCounter != expectedResolvers.size()) {
                 return false;
             }
```
Comment (volosied) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/TestJSFEL.war/src/com/ibm/ws/jsf22/el/components/CustomComponent.java:223]: Small thing:  I would move this size check before the list comparison.

File: dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/extprocessor/JSFExtensionFactory.java (hunk 3 above)
Comment (pnicolucci) [reply dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/extprocessor/JSFExtensionFactory.java:147]: Nope, the comment is correct, each comment talks about the other class so it is clear where the logic is taking place.

File: dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/ee/WASMyFacesContainerInitializer.java (hunk 2 above)
Comment (pnicolucci) [reply dev/io.openliberty.faces.4.0.internal/src/io/openliberty/faces40/internal/ee/WASMyFacesContainerInitializer.java:94]: Nope, the comment is correct, each comment talks about the other class so it is clear where the logic is taking place.

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/TestJSFEL.war/src/com/ibm/ws/jsf22/el/components/CustomComponent.java (hunk 4 above)
Comment (pnicolucci) [reply dev/com.ibm.ws.jsf.2.2_fat/test-applications/TestJSFEL.war/src/com/ibm/ws/jsf22/el/components/CustomComponent.java:223]: Yea, that makes sense let me get a change up for this. Thanks!

Comment (pnicolucci) [reply dev/com.ibm.ws.jsf.2.2_fat/test-applications/TestJSFEL.war/src/com/ibm/ws/jsf22/el/components/CustomComponent.java:223]: Need to compare and loop over the resolvers before we know the number of Resolvers found.

**Comment by pnicolucci** [comment]
#libby #build

**Review by volosied — APPROVED** [review]
Looks good. Merge when the back comes back :)

**Pull request merged**